File: app/src/main/java/io/github/marcocipriani01/telescopetouch/renderer/util/ColoredQuad.java
/*
 * Copyright 2021 Marco Cipriani (@marcocipriani01)
 *
 *  This program is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

package io.github.marcocipriani01.telescopetouch.renderer.util;

import javax.microedition.khronos.opengles.GL10;

public class ColoredQuad {
    private final float mR;
    private final float mG;
    private final float mB;
    private final float mA;
    private final VertexBuffer mPosition;

    public ColoredQuad(float r, float g, float b, float a,
                       float px, float py, float pz,
                       float ux, float uy, float uz,
                       float vx, float vy, float vz) {
        mPosition = new VertexBuffer(12);
        VertexBuffer vertexBuffer = mPosition;

        // Upper left
        vertexBuffer.addPoint(px - ux - vx, py - uy - vy, pz - uz - vz);

        // upper left
        vertexBuffer.addPoint(px - ux + vx, py - uy + vy, pz - uz + vz);

        // lower right
        vertexBuffer.addPoint(px + ux - vx, py + uy - vy, pz + uz - vz);

        // upper right
        vertexBuffer.addPoint(px + ux + vx, py + uy + vy, pz + uz + vz);

        mR = r;
        mG = g;
        mB = b;
        mA = a;
    }

    public void draw(GL10 gl) {
        gl.glEnableClientState(GL10.GL_VERTEX_ARRAY);
        gl.glDisableClientState(GL10.GL_TEXTURE_COORD_ARRAY);
        gl.glDisableClientState(GL10.GL_COLOR_ARRAY);

        // Enable blending if alpha != 1.
        if (mA != 1) {
            gl.glEnable(GL10.GL_BLEND);
            gl.glBlendFunc(GL10.GL_SRC_ALPHA, GL10.GL_ONE_MINUS_SRC_ALPHA);
        }

        gl.glDisable(GL10.GL_TEXTURE_2D);

        mPosition.set(gl);
        gl.glColor4f(mR, mG, mB, mA);

        gl.glDrawArrays(GL10.GL_TRIANGLE_STRIP, 0, 4);

        gl.glEnable(GL10.GL_TEXTURE_2D);

        // Disable blending if alpha != 1.
        if (mA != 1) {
            gl.glDisable(GL10.GL_BLEND);
        }
    }
}

File: app/src/main/java/com/google/android/stardroid/renderer/util/ColoredQuad.kt
// Copyright 2008 Google Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
package com.google.android.stardroid.renderer.util

import javax.microedition.khronos.opengles.GL10

class ColoredQuad(
    r: Float, g: Float, b: Float, a: Float,
    px: Float, py: Float, pz: Float,
    ux: Float, uy: Float, uz: Float,
    vx: Float, vy: Float, vz: Float
) {
    fun draw(gl: GL10) {
        gl.glEnableClientState(GL10.GL_VERTEX_ARRAY)
        gl.glDisableClientState(GL10.GL_TEXTURE_COORD_ARRAY)
        gl.glDisableClientState(GL10.GL_COLOR_ARRAY)

        // Enable blending if alpha != 1.
        if (mA != 1f) {
            gl.glEnable(GL10.GL_BLEND)
            gl.glBlendFunc(GL10.GL_SRC_ALPHA, GL10.GL_ONE_MINUS_SRC_ALPHA)
        }
        gl.glDisable(GL10.GL_TEXTURE_2D)
        mPosition!!.set(gl)
        gl.glColor4f(mR, mG, mB, mA)
        gl.glDrawArrays(GL10.GL_TRIANGLE_STRIP, 0, 4)
        gl.glEnable(GL10.GL_TEXTURE_2D)

        // Disable blending if alpha != 1.
        if (mA != 1f) {
            gl.glDisable(GL10.GL_BLEND)
        }
    }

    private var mPosition: VertexBuffer? = null
    private val mR: Float
    private val mG: Float
    private val mB: Float
    private val mA: Float

    init {
        mPosition = VertexBuffer(12)
        val vertexBuffer: VertexBuffer = mPosition as VertexBuffer

        // Upper left
        vertexBuffer.addPoint(px - ux - vx, py - uy - vy, pz - uz - vz)

        // upper left
        vertexBuffer.addPoint(px - ux + vx, py - uy + vy, pz - uz + vz)

        // lower right
        vertexBuffer.addPoint(px + ux - vx, py + uy - vy, pz + uz - vz)

        // upper right
        vertexBuffer.addPoint(px + ux + vx, py + uy + vy, pz + uz + vz)
        mR = r
        mG = g
        mB = b
        mA = a
    }
}